### ISSUE TITLE
clang incorrect order of static initialization with pch

### ISSUE THREAD

**Issue by zlojvavan** [llvm/llvm-project#67992]
first time clang user here so please bear with me. been using borland/codegear/embarcadero, msvc and gcc for years and recently decided to test clang just to discover incorrect order of static initializations in generated binary, but only with pch. msvc/gcc produce correct order with and without using of pch as well as clang without pch. I was able to find crazy workaround for my specific case and finally create working executable with clang and wonder whether it is known problem or not before trying to narrow down the bug in reduced test case.

Ubuntu clang version 17.0.2 (++20230929073225+098e653a5bed-1~exp1~20230929073241.48) here

basically, bug manifests itself as a class constructor running before its static inline members constructed BUT ONLY WITH PCH/clang combination.

`   #2 0x560ef541c218 in __cxx_global_var_init.17 /home/vavan/projects/lnsc/NetworkServer/lnsc/../../Common/helperCommonNew.h:13396:33
    #3 0x7f85a4629eba in call_init csu/../csu/libc-start.c:145:3
    #4 0x7f85a4629eba in __libc_start_main csu/../csu/libc-start.c:379:5
    #5 0x560ef5334024 in _start (/home/vavan/projects/lnsc/bin/x64/Debug/lnsc.out+0xdeb024) (BuildId: 1d6ddc43d7bed932)
`

**Comment by asl** [comment]
@zlojvavan Will you please attach a reproducible testcase?

**Comment by zlojvavan** [comment]
well, so if this is something new, I will have to try to create minimal reproducible testcase as at the moment I only have single real-world project built with clang with thousands of loc in problematic file. and as I was able to find my workaround this is of lower priority for me at the moment so not sure when exactly.

**Comment by zlojvavan** [comment]
ok, please see https://godbolt.org/z/r4vqqfP1K

**Comment by zlojvavan** [comment]
so, as it failed the same way with clang 18 I guess title is somewhat confusing and should be changed

**Comment by zlojvavan** [comment]
@Endilll is there anything else I should do to get rid of "incomplete" badge?

**Comment by Endilll** [comment]
No, CE link is an excellent reproducer. Thank you!

**Comment by shafik** [comment]
CC @ChuanqiXu9 maybe

**Comment by Endilll** [comment]
I haven't pinged him myself hoping that he's subscribed to `clang:PCH`, because I know for sure he is aware of this label.

**Comment by ChuanqiXu9** [comment]
Yeah, a reduced reproducer will be very helpful.

**Comment by zlojvavan** [comment]
@ChuanqiXu9 
> Yeah, a reduced reproducer will be very helpful.

do you want me to reduce it even more?

**Comment by ChuanqiXu9** [comment]
> https://godbolt.org/z/r4vqqfP1K

Oh, sorry for missing your comments.

**Comment by zlojvavan** [comment]
> sorry for missing your comments

no problem if you're ok with that godbolt example

**Comment by gregbedwell** [comment]
#91418 sounds like it could be another manifestation of the same issue?

**Comment by ChuanqiXu9** [comment]
Close this since this looks pretty similar with https://github.com/llvm/llvm-project/issues/91418.
